src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -492,47 +507,41 @@ void ObjectMonitor::enter_with_contention_mark(JavaThread *current, ObjectMonito
 
   freeze_result result;
 
-  { // Change java thread status to indicate blocked on monitor enter.
-    JavaThreadBlockedOnMonitorEnterState jtbmes(current, this);
+  assert(current->current_pending_monitor() == nullptr, "invariant");
```

Comment:
Question on this is pre-existent line. So if the callback from monitor_contended_enter calls Java code we might end up here again hitting this assert. In a product build it seems we would eventually hit SOE from recursive calls, I guess that’s the expected behavior?

Reply:
Thanks. This is a right observation. Yes, it is the expected behavior. The agent developers have to be careful in writing callbacks, especially because they can cause recursive calls.